backends/dpdk/dpdkArch.cpp

```diff
@@ -2448,4 +2446,86 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
     return m;
 }
 
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Type_Struct* s) {
+    auto fields = new IR::IndexedVector<IR::StructField>;
+    for (auto sf : s->fields) {
+        auto ftype = typeMap->getType(sf, true);
+        if (ftype->is<IR::Type_HeaderUnion>()) {
+            for (auto sfu : ftype->to<IR::Type_HeaderUnion>()->fields) {
+                cstring uName = sf->name.name + "_" + sfu->name.name;
+                auto uType = sfu->type->getP4Type();
+                fields->push_back(new IR::StructField(IR::ID(uName), uType));
+            }
+        } else {
+            fields->push_back(sf);
+        }
+    }
+    return new IR::Type_Struct(s->name, s->annotations, *fields);
```

Comment:
Instead of dereferencing fields, you could have created object and passed here. will it work?
IR::IndexedVector<IR::StructField> fields;

Reply:
Yes. it should work. will update the code